packages/@aws-cdk/cdk/lib/core/construct.ts

```diff
@@ -595,6 +603,53 @@ export class Construct implements IConstruct {
     this.node = new ConstructNode(this, scope, id);
   }
 
+  public get physicalName(): string {
+    if (this._physicalName === undefined) {
+      throw new Error(`Construct '${this.node.uniqueId}' has been used in a cross account and/or region fashion, ' +
+        'but does not have a physical name. Constructs have to have a physical name in order to be used cross account and/or region`);
+    } else {
+      return this._physicalName;
+    }
+  }
+
+  public get tryPhysicalName(): string | undefined {
+    return this._physicalName;
+  }
+
+  /**
+   * Assigns a physical name if the Construct does not have one already set.
+   *
+   * @param name the physical name to assign to this Construct.
+   *   If not provided, the name will be generated by the object
+   *   returned from {@link #physicalNameGenerator()}
+   * @returns true if the name was assigned
+   *   (that is, the Construct did not have a physical name set),
+   *   false otherwise
+   */
+  public assignPhysicalNameIfNotSet(name?: string): boolean {
+    if (this._physicalName) {
+      return false;
+    }
+
+    name = name || this.physicalNameGenerator().generate();
```

Comment:
I think this logic should moved somehow to the new `Token`.

Reply:
OK, I'll see If I can find an elegant way to make that happen.

Reply:
Done.